community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/commands/expressions/DistanceFunction.scala

```diff
@@ -26,19 +26,20 @@ import org.neo4j.cypher.internal.compatibility.v3_3.runtime.pipes.QueryState
 import org.neo4j.cypher.internal.compatibility.v3_3.runtime.{CRS, ExecutionContext}
 import org.neo4j.cypher.internal.frontend.v3_3.CypherTypeException
 import org.neo4j.graphdb.spatial.{Coordinate, Geometry, Point}
+import org.neo4j.values.AnyValue
+import org.neo4j.values.storable.{DoubleValue, Values}
+import org.neo4j.values.virtual.PointValue
 
 case class DistanceFunction(p1: Expression, p2: Expression) extends Expression {
 
   private val availableCalculators = Seq(HaversinCalculator, CartesianCalculator)
 
-  override def apply(ctx: ExecutionContext)(implicit state: QueryState): Any = {
+  override def apply(ctx: ExecutionContext)(implicit state: QueryState): AnyValue = {
     // TODO: Support better calculations, like https://en.wikipedia.org/wiki/Vincenty%27s_formulae
     // TODO: Support more coordinate systems
     (p1(ctx), p2(ctx)) match {
-      case (null, _) => null
-      case (_, null) => null
-      case (geometry1: Point, geometry2: Point) => calculateDistance(geometry1, geometry2)
-      case (geometry1: Geometry, geometry2: Geometry) => calculateDistance(geometry1, geometry2)
+      case (x, y) if x == Values.NO_VALUE || y == Values.NO_VALUE => Values.NO_VALUE
+      case (geometry1: PointValue, geometry2: PointValue) => calculateDistance(geometry1, geometry2)
```

Comment:
we do support geometry but we only have points so far. If a user pass in a geometry via parameters etc we will try to convert it to a point since that is the only geometry we know about. Same as before but we only convert/fail earlier.